### PR TITLE
job-ingest: validate jobspec using python/jsonschema

### PR DESCRIPTION
**Pull request by garlick**
This modifies `job-ingest` to spawn a python script to validate incoming jobspec, instead of calling into the C++ libjobspec library.   That library (which incidentally has been copied to flux-sched) is no longer needed in flux-core, so we get to drop the yaml-cpp build dependency.

The python script parses the jobspec schema once, then applies it multiple times to candidate jobspecs received on stdin, emitting results on stdout.  While considerably more flexible than the hardwired C++, the python/jsonschema script is slower.  For comparison, I timed submitting 100K jobs with `submitbench` (without munge signing):
* before this PR (use libjobspec inline with submit request): 23.45s
* one C++ external program: 31.6s
* one python script: 451s
* crew of 4 python scripts: 116s (speedup 3.9)
* crew of 8 python scripts: 59s (speedup 7.6)

Job submission probably won't be our main bottleneck, so I "tuned" the code that manages the submissions to spawn up to 4 python scripts on demand, and time them out after  5s of inactivity so there aren't a bunch of python interpreter instances hanging around on compute nodes looking menacing.  These tunings are currently hardwired.

The work crew uses libsubprocess to do the heavy lifting of process and stdio management, and that made it easy to get this started.  The interface to `job-ingest` is just a "request" function that returns a future.  It seems like eventually we might want to circle back and look at making this a generically useful "work crew class", so I did try to structure the code with that in mind, but probably that's a job for another time.

I left the C++ validator "coprocess" in the PR rather than squash it after libjobspec was removed, in case we might want to go back to it at some point.  It occurred to me that flux-sched could easily provide one of those if we wanted to outsource this work at some point.

Ultimately I think we will want a way to run a series of validators, possibly in parallel, or pass additional schemas to this python script to be applied serially to the preparsed (JSON) jobspec.  For now, I thought this was a reasonable chunk of work to call a PR.

### CODE REVIEW

**Comment by grondo** [comment]
This is so cool!

Before dropping yaml-cpp from our base docker images, we should be sure that this dependency is pulled in by flux-sched's Dockerfiles, and add it if not.

**Comment by garlick** [comment]
Another issue I ran into is I added the following to the valgrind sharness test and got some mem leaks in the broker's exec service (in the subprocess code).
```diff
diff --git a/t/valgrind/valgrind-workload.sh b/t/valgrind/valgrind-workload.sh
index 9523b759..fedd209b 100755
--- a/t/valgrind/valgrind-workload.sh
+++ b/t/valgrind/valgrind-workload.sh
@@ -7,3 +7,8 @@ echo FLUX_URI=$FLUX_URI
 for i in `seq 1 $NJOBS`; do
     flux wreckrun --ntasks $size /bin/true
 done
+
+# exercise job-ingest/job-management
+flux job submit <<EOT
+{"attributes":null,"tasks":[{"slot":"foo","count":{"per_slot":1},"command":"app"}],"version":1,"resources":[{"count":1,"type":"slot","with":[{"count":1,"type":"node"}],"label":"foo"}]}
+EOT
```
The leaks disappeared when I sent a `cmb.disconnect` message to the broker in `validate_destroy()`.  (Unlike `flux-exec` which is a process on the local connector, there is no automatic disconnect on module unload for `job-ingest`)

I guess that means I'm not cleaning up subprocesses properly.  What I'm doing is closing STDIN of the filter, then destroying the `flux_subprocess_t` without waiting for it, since I don't really have any way to wait for it.  @chu11 is there a way to do this?

**Comment by garlick** [comment]
Rebased on current master and added some commits to address some memory issues.

I added a refcount to futures, since the validation futures in this case are placed on a queue, and (in error cases) can be destroyed while still on the queue, triggering a use-after-free.  Now there's a refcount so that can't happen.

The I switched the valgrind sharness test over to use the new system.  (It submits 10 jobs and of course they don't run yet).  That uncovered some obvious memory leaks which are fixed.

Finally I added the above `cmb.disconnect` hack to request that the broker exec service clean up any leftover suprocess state from the ingest module.

**Comment by grondo** [comment]
> Finally I added the above cmb.disconnect hack to request that the broker exec service clean up any leftover suprocess state from the ingest module.

Is there an open issue on this? Maybe the subprocess api isn't quite right for this use case?

**Comment by garlick** [comment]
> Somewhere I had a (trivial) change to the valgrind tests that ran through a workload.d/*.sh directory instead of a static script. This allowed me to drop in tests for any subsystem (e.g. cron) to check for memleaks unrelated to a job workflow. Why don't we do something like that and then the wreckrun script could just be deleted when we actually remove the implementation?

Yes that sounds good.  I'll do that.

> Is there an open issue on this

Not yet, was hoping  that maybe I was missing something obvious, but even if so, maybe an issue is the best place to discuss.

**Comment by garlick** [comment]
Rebased on current master, squashed the valgrind test changes into one commit, squashed the two future changes into one commit, and fixed a merge error from the last push.

File: src/common/libflux/conf.c [hunk 1]
```diff
@@ -42,6 +42,10 @@ static struct config default_config[] = {
     { "no_docs_path",   INSTALLED_NO_DOCS_PATH,     INTREE_NO_DOCS_PATH },
     { "rundir",         INSTALLED_RUNDIR,           NULL },
     { "bindir",         INSTALLED_BINDIR,           INTREE_BINDIR },
+    { "jobspec_validate_path", INSTALLED_JOBSPEC_VALIDATE_PATH,
+                                            INTREE_JOBSPEC_VALIDATE_PATH },
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libflux/conf.c]: Actually, nevermind looking on a console I can see this is correct. Can we just pretend this comment was never made?

[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libflux/conf.c]: OK by me.

File: src/modules/job-ingest/Makefile.am [hunk 2]
```diff
@@ -23,9 +30,8 @@ job_ingest_la_LIBADD = $(fluxmod_libadd) \
...
-job_ingest_la_SOURCES += jobspec_wrap.cpp jobspec.h
-job_ingest_la_LIBADD += \
-		    $(top_builddir)/src/common/libjobspec/libjobspec.la \
-                    $(YAMLCPP_LIBS)
-endif
+dist_fluxlibexec_SCRIPTS = \
+	validators/validate-schema.py
+
+fluxschemadir = $(datadir)/flux/
+dist_fluxschema_DATA = schemas/jobspec.jsonschema
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/modules/job-ingest/Makefile.am]: Sure, I'll do that.

**Comment by garlick** [comment]
Right, I think the validator is probably in an early state here and there may be things we can do to improve error messages down the road.

The first thing I would like to do is look into the fairly sophisticated [exception model](https://python-jsonschema.readthedocs.io/en/latest/errors/)  for `jsonschema.validate()`.  There is a stack of exceptions and we are only returning a summary (or maybe the top of the stack?)

Then perhaps investigate whether some of the jobspec schema might need to be restructured to avoid certain constructs that make useful errors elusive.  For example, the [oneOf]( https://json-schema.org/understanding-json-schema/reference/combining.html) operator, tends to result in "none of the available schemas could validate" or similar, rather than anything concrete.

Finally, it would be pretty easy to short circuit some common errors using hardwired python code, like the version one with the comically unhelpful message you quoted.

But I think maybe we are OK for now, especially since for the near term we'll be generating jobspec from simple commands.

**Comment by SteVwonder** [comment]
> python unicode strings from jsonschema exceptions are being converted to JSON strings with the leading u included

Is this happening with python 2, 3, or both?  I suspect it doesn't happen in 2 but does in 3.  Can you post the contents of broke.json as a reproducer?

File: src/modules/job-ingest/validators/validate-schema.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+try:
+    with open(args.schema) as fd:
+        schema = json.load(fd)
+except (OSError, IOError) as e:
+    sys.exit(args.schema + ": " + e.strerror)
+except ValueError as e:
+    sys.exit(args.schema + ": " + str(e))
+
+sys.stderr.write("ready\n")
```
Comment (SteVwonder) [inline src/modules/job-ingest/validators/validate-schema.py]: For printing to stderr in python, I recommend the following:
```python
from __future__ import print_function

print("ready", file=sys.stderr)
```

**Comment by garlick** [comment]
Looks like v2 does it:
```
$ python --version
Python 2.7.15rc1
$ cd src/modules/job-ingest
$ echo '{"version":2}' | validators/validate-schema.py -s schemas/jobspec.jsonschema
{"errnum":1,"errstr":"u'resources' is a required property"}
```
If I change the shebang to `#/usr/bin/env python3` the problem goes away.

File: src/modules/job-ingest/validators/validate-schema.py (hunk 3 above)
Comment (garlick) [reply src/modules/job-ingest/validators/validate-schema.py]: OK, thanks!

Comment (grondo) [reply src/modules/job-ingest/validators/validate-schema.py]: Is the `from __future__` a python 2/3 compatibility thing?

**Comment by garlick** [comment]
Just forced a push after squashing the incremental bug fixes and  installed schema path change suggested by @grondo.

@SteVwonder's suggestion for validate-schema.py was added but not squashed.

**Comment by garlick** [comment]
Fixed up the work crew load management a bit.  I can squash the rest of this down if we think it's getting close.

**Comment by garlick** [comment]
I went ahead and squashed and forced a push.

File: src/modules/job-ingest/validators/validate-schema.py (hunk 3 above)
Comment (SteVwonder) [reply src/modules/job-ingest/validators/validate-schema.py]: > Is the from __future__ a python 2/3 compatibility thing?

Yep!  In py2, print is a statement, while in py3 it is a function.  The print function is what add the support for the `file` keyword arg. This pulls in the backported print function into py2 (and is a no-op is py3).

**Comment by SteVwonder** [comment]
> python unicode strings from jsonschema exceptions are being converted to JSON strings with the leading u included

Thanks for the reproducer @garlick.  The error message coming out of jsonschema is already encoded into a str, so the easy fix I was hoping for isn't doable.  There is a thread on jsonschema repo about this problem: https://github.com/Julian/jsonschema/issues/243.  The dev has decided to focus on explicit, dev-focused error messages rather than user-friendly error messages. So the u"string" is considered a feature not a bug. 

Conveniently though, someone has already a made a pretty printer for jsonschema errors that should fix this problem: https://github.com/ccpgames/jsonschema-errorprinter

> but some of the errors returned are a bit vague, such as when I tried to submit a jobspec with "version": 2:

The above pretty printer could potentially solve this problem as well by giving context to the vague error message.

**Comment by garlick** [comment]
Thanks!  They are just substituting out the `u'` so maybe I can just do that?  (seems to work)

```diff
diff --git a/src/modules/job-ingest/validators/validate-schema.py b/src/modules/job-ingest/validators/validate-schema.py
index 7a1047a0..9e979da3 100755
--- a/src/modules/job-ingest/validators/validate-schema.py
+++ b/src/modules/job-ingest/validators/validate-schema.py
@@ -31,7 +31,7 @@ def validate(schema, line):
     except ValueError as e:
         return (1, str(e))
     except jsonschema.exceptions.ValidationError as e:
-        return (1, e.message)
+        return (1, e.message.replace("u'", "'"))
     return (0, None)
 ```

**Comment by garlick** [comment]
I'm still a little wary of creating a slough of python package dependencies, so maybe it makes sense to hold off on the pretty printer and just make the above change for now?

**Comment by garlick** [comment]
Incorporated that `validate-jobspec.py` output substitution change, squashed, and rebased on current master.

**Comment by garlick** [comment]
Rebased.

**Comment by SteVwonder** [comment]
Thanks @garlick for all the work on this.  The worker pool is really cool!

> They are just substituting out the u' so maybe I can just do that?

Oh! Nice catch. I didn't look closely at their solution.  It's convenient that almost no english words end in u 😆 

> I'm still a little wary of creating a slough of python package dependencies

Agreed! For the pretty printer, I don't think it is even in any package managers, not even pip. So if we do eventually go that route, vendoring would probably be the way to go.

**Comment by garlick** [comment]
Thanks everybody!

**Pull request merged**